File: keris-academic-search/scripts/keris_academic.py
#!/usr/bin/env python3
"""Read-only KERIS/RISS academic metadata search helper using stdlib only."""

from __future__ import annotations

import argparse
import json
import os
import pathlib
import sys
import urllib.error
import urllib.parse
import urllib.request
import xml.etree.ElementTree as ET
from typing import Any, Dict, List, Optional

DEFAULT_PROXY_BASE_URL = "https://k-skill-proxy.nomadamas.org"
DEFAULT_SECRETS_PATH = pathlib.Path("~/.config/k-skill/secrets.env").expanduser()
RISS_OPEN_API_URL = "https://www.riss.kr/openApi"
RESOURCE_TYPE_MAP = {"ALL": ["T", "A", "O", "U", "F", "S"], "T": ["T"], "A": ["A", "O"], "D": ["A"], "B": ["U"]}
PROXY_DOWN_MSG = "설정된 k-skill-proxy 프록시 서버가 응답하지 않습니다. 잠시 후 재시도하거나 운영자에게 문의하세요."
PROXY_NOT_CONFIGURED_MSG = "k-skill-proxy에 RISS API 키가 설정되어 있지 않습니다. 운영자에게 문의하세요."


class HelperError(RuntimeError):
    pass


def load_secrets(path: pathlib.Path) -> Dict[str, str]:
    try:
        lines = path.read_text(encoding="utf-8").splitlines()
    except OSError:
        return {}
    values = {}
    for raw in lines:
        line = raw.strip()
        if line and not line.startswith("#") and "=" in line:
            key, value = line.split("=", 1)
            values[key.strip()] = value.strip().strip('"').strip("'")
    return values


def resolve_api_key(args: argparse.Namespace) -> Optional[str]:
    for name in ("KSKILL_RISS_API_KEY", "RISS_API_KEY"):
        value = os.environ.get(name)
        if value and value.strip():
            return value.strip()
    secrets = load_secrets(pathlib.Path(args.secrets_path).expanduser())
    value = secrets.get("KSKILL_RISS_API_KEY") or secrets.get("RISS_API_KEY")
    return value.strip() if value and value.strip() else None


def build_query(args: argparse.Namespace) -> Dict[str, Any]:
    query = {}
    for field in ("keyword", "title", "author", "subject", "publisher"):
        value = getattr(args, field)
        if value and value.strip():
            text = value.strip()
            if len(text) > 200 or any(ord(char) < 32 or ord(char) == 127 for char in text):
                raise HelperError(f"{field} 값은 1~200자의 일반 텍스트여야 합니다.")
            query[field] = text
    if not query:
        raise HelperError("--keyword, --title, --author, --subject, --publisher 중 하나 이상을 입력하세요.")
    if not 1 <= args.page <= 100000:
        raise HelperError("page 값은 1~100000 범위여야 합니다.")
    if not 1 <= args.page_size <= 100:
        raise HelperError("pageSize 값은 1~100 범위여야 합니다.")
    if len(RESOURCE_TYPE_MAP[args.resource_type]) > 1 and args.page > 1:
        raise HelperError("Combined resourceType searches support page 1 only; later pages require a single resource type.")
    query.update({"resourceType": args.resource_type, "page": args.page, "pageSize": args.page_size})
    return query


def build_url(args: argparse.Namespace, query: Dict[str, Any], api_key: Optional[str]) -> str:
    del api_key
    return f"{args.proxy_base_url.rstrip('/')}/v1/keris-academic/search?{urllib.parse.urlencode(query)}"


def build_direct_urls(args: argparse.Namespace, query: Dict[str, Any], api_key: Optional[str]) -> List[str]:
    if not api_key:
        raise HelperError("KSKILL_RISS_API_KEY 또는 RISS_API_KEY가 없습니다. RISS API 센터에서 검색 API 키를 발급받아 설정하세요.")
    rsnum = ((query["page"] - 1) * query["pageSize"]) + 1
    urls = []
    for upstream_type in RESOURCE_TYPE_MAP[query["resourceType"]]:
        params = {key: value for key, value in query.items() if key not in {"resourceType", "page", "pageSize"}}
        params.update({"key": api_key, "version": "1.0", "type": upstream_type, "rsnum": rsnum, "rowcount": query["pageSize"]})
        urls.append(f"{RISS_OPEN_API_URL}?{urllib.parse.urlencode(params)}")
    return urls


def http_get_json(url: str, timeout: int, via_proxy: bool = True) -> Dict[str, Any]:
    request = urllib.request.Request(url, headers={"accept": "application/json", "user-agent": "k-skill/keris-academic-search"})
    try:
        with urllib.request.urlopen(request, timeout=timeout) as response:
            raw = response.read().decode("utf-8", errors="replace")
    except urllib.error.HTTPError as error:
        body = error.read().decode("utf-8", errors="replace") if error.fp else ""
        try:
            payload = json.loads(body)
        except json.JSONDecodeError:
            payload = {}
        if via_proxy and error.code == 503 and payload.get("error") == "upstream_not_configured":
            raise HelperError(PROXY_NOT_CONFIGURED_MSG) from error
        raise HelperError(str(payload.get("message") or f"API HTTP 오류: {error.code} {error.reason}")) from error
    except urllib.error.URLError as error:
        raise HelperError(f"{PROXY_DOWN_MSG} (상세: {error.reason})" if via_proxy else f"RISS API 네트워크 오류: {error.reason}") from error
    except TimeoutError as error:
        target = "프록시 서버" if via_proxy else "RISS API"
        raise HelperError(f"{target} 요청 시간이 초과되었습니다.") from error
    try:
        payload = json.loads(raw)
    except json.JSONDecodeError as error:
        raise HelperError("API 응답이 올바른 JSON이 아닙니다.") from error
    if not isinstance(payload, dict):
        raise HelperError("API 응답 형식이 올바르지 않습니다.")
    return payload


def parse_riss_xml(raw: bytes) -> Dict[str, Any]:
    try:
        root = ET.fromstring(raw)
    except ET.ParseError as error:
        raise HelperError("RISS API 응답 XML이 올바르지 않습니다.") from error
    if root.tag != "record":
        raise HelperError("RISS API 응답 envelope가 올바르지 않습니다.")
    error_text = root.findtext("./head/Error")
    if error_text is None or not error_text.strip():
        raise HelperError("RISS API 응답에 상태 메타데이터가 없습니다.")
    code = error_text.strip()
    message = (root.findtext("./head/ErrorMessage") or "Unknown RISS error").strip()
    if code not in {"0", "000"}:
        raise HelperError(f"RISS API 오류 {code}: {message}")
    items = []
    for metadata in root.findall("./metadata"):
        values: Dict[str, List[str]] = {}
        for child in metadata:
            value = "".join(child.itertext()).strip()
            if value:
                values.setdefault(child.tag.removeprefix("riss.").lower(), []).append(value)
        first = lambda key: values.get(key, [None])[0]
        image, charge = (first("image") or "").upper(), first("charge")
        available = True if image == "Y" else False if image == "N" else None
        access = "free" if available and charge == "1" else "paid_or_restricted" if available and charge == "0" else "available" if available else "none" if available is False else "unknown"
        author = first("author") or ""
        authors = [part.strip() for part in author.replace(";", ",").split(",") if part.strip()] or ([author] if author else [])
        items.append({"resource_type": first("type"), "title": first("title"), "authors": authors, "publisher": first("publisher"), "year": first("pubdate"), "publication": first("stitle"), "material_type": first("mtype"), "link": first("url"), "full_text_available": available, "full_text_access": access, "holdings": values.get("holdings", [])})
    total_text = root.findtext("./head/totalcount")
    try:
        total_count = int(total_text) if total_text is not None else len(items)
    except ValueError as error:
        raise HelperError("RISS API 응답 totalcount가 올바른 정수가 아닙니다.") from error
    if total_count < 0:
        raise HelperError("RISS API 응답 totalcount가 음수입니다.")
    return {"total_count": total_count, "items": items}


def http_get_direct_xml(urls: List[str], timeout: int, page: int, page_size: int) -> Dict[str, Any]:
    results = []
    for url in urls:
        request = urllib.request.Request(url, headers={"accept": "application/xml", "user-agent": "k-skill/keris-academic-search"})
        try:
            with urllib.request.urlopen(request, timeout=timeout) as response:
                results.append(parse_riss_xml(response.read()))
        except (urllib.error.HTTPError, urllib.error.URLError) as error:
            raise HelperError(f"RISS API 요청 실패: {error}") from error
        except TimeoutError as error:
            raise HelperError("RISS API 요청 시간이 초과되었습니다.") from error
    queues = [list(result["items"]) for result in results]
    items = []
    while len(items) < page_size and any(queues):
        for queue in queues:
            if queue:
                items.append(queue.pop(0))
            if len(items) >= page_size:
                break
    return {"page": page, "page_size": page_size, "total_count": sum(result["total_count"] for result in results), "items": items}


def format_text(payload: Dict[str, Any]) -> str:
    items = payload.get("items") or []
    lines = [f"RISS 학술자료 검색 결과: {payload.get('total_count', len(items))}건"]
    labels = {"free": "원문 있음(무료 표시)", "paid_or_restricted": "원문 있음(유료/기관권한 가능)", "available": "원문 있음(접근조건 확인 필요)", "none": "원문 없음", "unknown": "원문 여부 미상"}
    for item in items:
        lines.append(f"- {item.get('title') or '제목 없음'} / {', '.join(item.get('authors') or []) or '저자 미상'} / {item.get('publisher') or item.get('publication') or '발행처 미상'} / {item.get('year') or '연도 미상'} / {labels.get(item.get('full_text_access'), '원문 여부 미상')}")
        if item.get("link"):
            lines.append(f"  {item['link']}")
    if not items:
        lines.append("조건에 맞는 학술자료가 없습니다.")
    return "\n".join(lines)


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="KERIS/RISS 학술자료 메타데이터 검색")
    search = parser.add_subparsers(dest="command", required=True).add_parser("search")
    for field in ("keyword", "title", "author", "subject", "publisher"):
        search.add_argument(f"--{field}")
    search.add_argument("--resource-type", choices=RESOURCE_TYPE_MAP, default="ALL")
    search.add_argument("--page", type=int, default=1)
    search.add_argument("--page-size", type=int, default=10)
    search.add_argument("--proxy-base-url", default=os.environ.get("KSKILL_PROXY_BASE_URL", DEFAULT_PROXY_BASE_URL))
    search.add_argument("--direct", action="store_true")
    search.add_argument("--secrets-path", default=str(DEFAULT_SECRETS_PATH))
    search.add_argument("--timeout", type=int, default=20)
    search.add_argument("--dry-run", action="store_true")
    search.add_argument("--json", action="store_true")
    return parser


def parse_args(argv: Optional[List[str]] = None) -> argparse.Namespace:
    return build_parser().parse_args(argv)


def run(argv: Optional[List[str]] = None) -> int:
    try:
        args, query = parse_args(argv), None
        query = build_query(args)
        if args.direct:
            urls = build_direct_urls(args, query, "REDACTED" if args.dry_run else resolve_api_key(args))
            if args.dry_run:
                print(json.dumps({"operation": "search", "urls": urls, "query": query}, ensure_ascii=False, indent=2)); return 0
            payload = http_get_direct_xml(urls, args.timeout, query["page"], query["pageSize"])
        else:
            url = build_url(args, query, None)
            if args.dry_run:
                print(json.dumps({"operation": "search", "url": url, "query": query}, ensure_ascii=False, indent=2)); return 0
            payload = http_get_json(url, args.timeout)
        print(json.dumps(payload, ensure_ascii=False, indent=2) if args.json else format_text(payload)); return 0
    except HelperError as error:
        print(str(error), file=sys.stderr); return 1


if __name__ == "__main__":
    raise SystemExit(run())

File: packages/k-skill-proxy/src/keris-academic.js
const RISS_OPEN_API_URL = "https://www.riss.kr/openApi";
const RESOURCE_TYPE_MAP = Object.freeze({
  ALL: ["T", "A", "O", "U", "F", "S"],
  T: ["T"],
  A: ["A", "O"],
  D: ["A"],
  B: ["U"]
});
const SEARCH_FIELDS = ["keyword", "title", "author", "subject", "publisher"];

function trimOrNull(value) {
  if (value === undefined || value === null) return null;
  const text = String(value).trim();
  return text || null;
}

function parseBoundedInteger(value, { defaultValue, max, label }) {
  const text = trimOrNull(value);
  if (text === null) return defaultValue;
  if (!/^\d+$/.test(text)) throw new Error(`${label} must be an integer.`);
  const parsed = Number.parseInt(text, 10);
  if (parsed < 1 || parsed > max) throw new Error(`${label} must be between 1 and ${max}.`);
  return parsed;
}

function validateSearchText(value, label) {
  const text = trimOrNull(value);
  if (text === null) return null;
  if (text.length > 200 || /[\u0000-\u001f\u007f]/.test(text)) {
    throw new Error(`Provide valid ${label} (1-200 characters).`);
  }
  return text;
}

function rejectUnsupportedQuery(query) {
  const supported = new Set([
    ...SEARCH_FIELDS,
    "resourceType", "resource_type", "type", "page", "pageSize", "page_size"
  ]);
  const controlled = new Set(["key", "servicekey", "version", "rsnum", "rowcount"]);
  for (const key of Object.keys(query)) {
    if (controlled.has(key.toLowerCase())) throw new Error(`${key} is controlled by the proxy server.`);
    if (!supported.has(key)) throw new Error(`${key} is not supported for KERIS academic search.`);
  }
}

function normalizeKerisAcademicQuery(query = {}) {
  rejectUnsupportedQuery(query);
  const normalized = {};
  for (const field of SEARCH_FIELDS) {
    const value = validateSearchText(query[field], field);
    if (value !== null) normalized[field] = value;
  }
  if (!SEARCH_FIELDS.some((field) => normalized[field])) {
    throw new Error("Provide at least one search field: keyword, title, author, subject, or publisher.");
  }
  const resourceType = (trimOrNull(query.resourceType ?? query.resource_type ?? query.type) || "ALL").toUpperCase();
  const upstreamTypes = RESOURCE_TYPE_MAP[resourceType];
  if (!upstreamTypes) throw new Error(`resourceType must be one of: ${Object.keys(RESOURCE_TYPE_MAP).join(", ")}.`);
  const page = parseBoundedInteger(query.page, { defaultValue: 1, max: 100000, label: "page" });
  const pageSize = parseBoundedInteger(query.pageSize ?? query.page_size, {
    defaultValue: 10,
    max: 100,
    label: "pageSize"
  });
  if (upstreamTypes.length > 1 && page > 1) {
    throw new Error("Combined resourceType searches support page 1 only; choose a single type for later pages.");
  }
  return {
    ...normalized,
    resourceType,
    upstreamTypes: [...upstreamTypes],
    page,
    pageSize,
    rsnum: ((page - 1) * pageSize) + 1
  };
}

function decodeXmlEntities(value) {
  return value.replace(/&(#x[0-9a-f]+|#\d+|amp|lt|gt|quot|apos);/gi, (match, entity) => {
    const lower = entity.toLowerCase();
    if (lower === "amp") return "&";
    if (lower === "lt") return "<";
    if (lower === "gt") return ">";
    if (lower === "quot") return '"';
    if (lower === "apos") return "'";
    const codePoint = lower.startsWith("#x")
      ? Number.parseInt(lower.slice(2), 16)
      : Number.parseInt(lower.slice(1), 10);
    return Number.isFinite(codePoint) ? String.fromCodePoint(codePoint) : match;
  });
}

function parseXmlDocument(xml) {
  const source = trimOrNull(xml);
  if (!source || !source.startsWith("<")) throw new Error("RISS upstream did not return XML.");
  if (/<!DOCTYPE|<!ENTITY/i.test(source)) throw new Error("RISS upstream XML contains unsupported declarations.");
  const document = { name: "#document", children: [], text: "" };
  const stack = [document];
  const tokens = source.match(/<!--[^]*?-->|<!\[CDATA\[[^]*?\]\]>|<\?[^]*?\?>|<[^>]+>|[^<]+/g);
  if (!tokens || tokens.join("") !== source) throw new Error("RISS upstream returned malformed XML.");
  for (const token of tokens) {
    if (token.startsWith("<!--") || token.startsWith("<?")) continue;
    if (token.startsWith("<![CDATA[")) {
      stack[stack.length - 1].text += token.slice(9, -3);
    } else if (!token.startsWith("<")) {
      stack[stack.length - 1].text += decodeXmlEntities(token);
    } else if (token.startsWith("</")) {
      const name = token.slice(2, -1).trim();
      if (stack.length === 1 || stack[stack.length - 1].name !== name) throw new Error("RISS upstream returned malformed XML.");
      stack.pop();
    } else {
      if (token.startsWith("<!")) throw new Error("RISS upstream XML contains unsupported markup.");
      const selfClosing = /\/\s*>$/.test(token);
      const match = token.match(/^<([A-Za-z_][\w:.-]*)(?:\s[^>]*)?\/?\s*>$/);
      if (!match) throw new Error("RISS upstream returned malformed XML.");
      const node = { name: match[1], children: [], text: "" };
      stack[stack.length - 1].children.push(node);
      if (!selfClosing) stack.push(node);
    }
  }
  if (stack.length !== 1 || document.children.length !== 1) throw new Error("RISS upstream returned malformed XML.");
  return document.children[0];
}

function child(node, name) {
  return node.children.find((entry) => entry.name.toLowerCase() === name.toLowerCase()) || null;
}

function children(node, name) {
  return node.children.filter((entry) => entry.name.toLowerCase() === name.toLowerCase());
}

function nodeText(node) {
  if (!node) return null;
  const nested = node.children.map(nodeText).filter(Boolean).join(" ");
  return trimOrNull(`${node.text}${nested ? ` ${nested}` : ""}`);
}

function metadataValues(node) {
  const values = new Map();
  for (const entry of node.children) {
    const key = entry.name.replace(/^riss\./i, "").toLowerCase();
    const value = nodeText(entry);
    if (value === null) continue;
    if (!values.has(key)) values.set(key, []);
    values.get(key).push(value);
  }
  return values;
}

function first(values, key) {
  return values.get(key)?.[0] ?? null;
}

function splitAuthors(value) {
  if (!value) return [];
  const explicit = value.split(/\s*[;,|]\s*/).filter(Boolean);
  if (explicit.length > 1) return explicit;
  const koreanNames = value.split(/\s+/).filter(Boolean);
  if (koreanNames.length > 1 && koreanNames.every((name) => /^[가-힣]{2,4}$/.test(name))) return koreanNames;
  return [value];
}

function normalizeMetadata(node) {
  const values = metadataValues(node);
  const image = first(values, "image")?.toUpperCase() ?? null;
  const charge = first(values, "charge");
  const fullTextAvailable = image === "Y" ? true : image === "N" ? false : null;
  let fullTextAccess = "unknown";
  if (fullTextAvailable === false) fullTextAccess = "none";
  if (fullTextAvailable === true && charge === "1") fullTextAccess = "free";
  if (fullTextAvailable === true && charge === "0") fullTextAccess = "paid_or_restricted";
  if (fullTextAvailable === true && charge === null) fullTextAccess = "available";
  return {
    resource_type: first(values, "type"),
    title: first(values, "title"),
    authors: splitAuthors(first(values, "author")),
    publisher: first(values, "publisher"),
    year: first(values, "pubdate"),
    publication: first(values, "stitle"),
    material_type: first(values, "mtype"),
    link: first(values, "url"),
    full_text_available: fullTextAvailable,
    full_text_access: fullTextAccess,
    holdings: values.get("holdings") || [],
  };
}

function parseRissXml(xml) {
  let root;
  try {
    root = parseXmlDocument(xml);
  } catch (error) {
    error.code = "upstream_invalid_response";
    throw error;
  }
  if (root.name.toLowerCase() !== "record") {
    const error = new Error("RISS upstream returned an invalid response envelope.");
    error.code = "upstream_invalid_response";
    throw error;
  }
  const head = child(root, "head");
  if (!head) {
    const error = new Error("RISS upstream response is missing head metadata.");
    error.code = "upstream_invalid_response";
    throw error;
  }
  const errorCode = nodeText(child(head, "Error"));
  if (errorCode === null) {
    const error = new Error("RISS upstream response is missing status metadata.");
    error.code = "upstream_invalid_response";
    throw error;
  }
  const errorMessage = nodeText(child(head, "ErrorMessage")) || "Unknown RISS error";
  if (errorCode !== null && !["0", "000"].includes(errorCode)) {
    const error = new Error(errorMessage);
    error.code = ["004", "4"].includes(errorCode) || /인증|AUTH|KEY/i.test(errorMessage)
      ? "upstream_forbidden"
      : /쿼터|호출량|초과|QUOTA|LIMIT/i.test(errorMessage)
        ? "upstream_quota_exceeded"
        : "upstream_error";
    error.upstreamCode = errorCode;
    throw error;
  }
  const items = children(root, "metadata").map(normalizeMetadata);
  const totalCount = Number.parseInt(nodeText(child(head, "totalcount")) ?? String(items.length), 10);
  if (!Number.isFinite(totalCount) || totalCount < 0) {
    const error = new Error("RISS upstream returned invalid totalcount metadata.");
    error.code = "upstream_invalid_response";
    throw error;
  }
  return { totalCount, items };
}

function errorResult(error, message) {
  return { status_code: 502, error, message };
}

async function fetchOneRissType({ params, upstreamType, apiKey, fetchImpl }) {
  const url = new URL(RISS_OPEN_API_URL);
  url.searchParams.set("key", apiKey);
  url.searchParams.set("version", "1.0");
  url.searchParams.set("type", upstreamType);
  for (const field of SEARCH_FIELDS) if (params[field]) url.searchParams.set(field, params[field]);
  url.searchParams.set("rsnum", String(params.rsnum));
  url.searchParams.set("rowcount", String(params.pageSize));
  let response;
  try {
    response = await fetchImpl(url.toString(), { signal: AbortSignal.timeout(20000) });
  } catch {
    return errorResult("upstream_unavailable", "RISS upstream request failed.");
  }
  const text = await response.text();
  if (response.status === 401 || response.status === 403) return errorResult("upstream_forbidden", "RISS upstream rejected the proxy key.");
  if (response.status === 429) return errorResult("upstream_quota_exceeded", "RISS upstream quota was exceeded.");
  if (!response.ok) return errorResult("upstream_error", `RISS upstream returned HTTP ${response.status}.`);
  if (!text.trim()) return errorResult("upstream_invalid_response", "RISS upstream returned an empty response.");
  try {
    return parseRissXml(text);
  } catch (error) {
    return errorResult(error.code || "upstream_invalid_response", `RISS upstream error response: ${error.message}`);
  }
}

async function fetchKerisAcademicSearch({ params, apiKey, fetchImpl = global.fetch }) {
  if (!apiKey) {
    return {
      status_code: 503,
      error: "upstream_not_configured",
      message: "KSKILL_RISS_API_KEY is not configured on the proxy server. RISS_API_KEY is accepted as a compatibility fallback."
    };
  }
  const results = await Promise.all(params.upstreamTypes.map((upstreamType) => fetchOneRissType({
    params, upstreamType, apiKey, fetchImpl
  })));
  const failed = results.find((result) => result.error);
  if (failed) return failed;
  const query = Object.fromEntries(Object.entries(params).filter(([key]) => !new Set(["upstreamTypes", "rsnum"]).has(key)));
  const queues = results.map((result) => [...result.items]);
  const items = [];
  while (items.length < params.pageSize && queues.some((queue) => queue.length > 0)) {
    for (const queue of queues) {
      if (queue.length > 0) items.push(queue.shift());
      if (items.length >= params.pageSize) break;
    }
  }
  return {
    query,
    page: params.page,
    page_size: params.pageSize,
    total_count: results.reduce((sum, result) => sum + result.totalCount, 0),
    items,
    source: {
      provider: "KERIS RISS Open API",
      upstream: RISS_OPEN_API_URL,
      upstream_types: params.upstreamTypes,
      response_format: "XML",
      data_go_kr_dataset: null,
      related_catalog_dataset: "15071949"
    }
  };
}

module.exports = {
  RESOURCE_TYPE_MAP,
  RISS_OPEN_API_URL,
  fetchKerisAcademicSearch,
  normalizeKerisAcademicQuery,
  parseRissXml
};
